packages/affiliates/libs/affiliates/dappmining_v2.js

```diff
@@ -0,0 +1,278 @@
+const assert = require("assert");
+const highland = require("highland");
+const moment = require("moment");
+
+const { DecodeTransaction, DecodeLog } = require("../contracts");
+const { EmpAttributions, EmpBalances } = require("../processors");
+const { AttributionLookback, Balances } = require("../models");
+
+// Module will calculate the Dapp Mining rewards for a single EMP
+module.exports = ({ queries, empAbi, web3 }) => {
+  assert(queries, "requires queries bigquery interface");
+  assert(empAbi, "requires empAbi");
+  assert(web3, "requires web3");
+
+  const { toBN, fromWei } = web3.utils;
+
+  function getEventStream(empAddress, start, end) {
+    const stream = queries.streamLogsByContract(empAddress, start, end);
+    const decode = DecodeLog(empAbi);
+    return highland(stream).map(log => {
+      return decode(log, {
+        blockNumber: log.block_number,
+        blockTimestamp: moment(log.block_timestamp.value).valueOf(),
+        ...log
+      });
+    });
+  }
+
+  function getAttributionStream(empAddress, start, end) {
+    // We get all attributions from the trace stream. This is very similar to transactions query but
+    // transactions do not contain transactions called from other contracts! Trace does.
+    const stream = queries.streamTracesByContract(empAddress, start, end);
+    const decodeTx = DecodeTransaction(empAbi);
+
+    return (
+      highland(stream)
+        // trace transactions may contain the contract creation event, we cant decode this so filter out
+        // fyi non contract creation "trace_type" are called "call"
+        .filter(tx => tx.trace_type !== "create" && tx.error == null)
+        .map(tx => {
+          return decodeTx(tx, {
+            blockNumber: tx.block_number,
+            blockTimestamp: moment(tx.block_timestamp.value).valueOf(),
+            ...tx
+          });
+        })
+        // this filters only emp contract "create" calls. We could technically look at more, but these
+        // would be the only "attributed" calls as per the spec
+        .filter(tx => tx.name == "create")
+    );
+  }
+
+  // returns the proportion of rewards given to each developer for a specific point in time
+  function calculateBlockReward(
+    { attributions, balances, whitelist },
+    // return all tags with a percent value for rewards
+    rewards = {}
+  ) {
+    // for every user with a token balance
+    balances.keys().forEach(userid => {
+      // get users balance
+      const balance = balances.get(userid);
+      const attributionPercents = attributions.getAttributionPercents(userid, balance);
```

Comment:
this function now returns an object keyed by developer addresses that show attribution % based on the users balance.

Reply:
+1